Add ECC curve selection order config in TLS server

ferd opened the pull request:
As per RFC 4492 Sec 5.1, the preferred order of selection of named
curves is based on client preferences.

Currently, the SSL application only picks entries according to the
absolute order of entries as tracked in a hardcoded list in code.

This patch changes things so that the client-specified order is
preferred. It also allows a mode where the server can be configured to
override the client's preferred order with its own, although the chosen
ECC must still be within both lists.

Currently, this PR still lacks tests. As seen in https://github.com/erlang/otp/pull/1179, there are no good
tests for this feature as of today, so we'll have to figure that one out.

For now, I've run the existing tests with the default config to confirm
nothing broke, and ran additional manual tests by inspecting SSL packets
by hand to see that curve selection does indeed take place:
---

Default config (equivalent to client preference):

Server:

```
1> BaseConfig = [{cacertfile, "/home/ferd/code/self/otp/lib/ssl/cacerts.pem"}, {certfile, "/home/ferd/code/self/otp/lib/ssl/cert.pem"}, {keyfile, "/home/ferd/code/self/otp/lib/ssl/key.pem"}, {reuseaddr, true}, {active,false}], 
2> f(F), F= fun() -> {ok,L} = ssl:listen(9234, BaseConfig), 
2> {ok,S} = ssl:transport_accept(L), ok = ssl:ssl_accept(S),
2> {ok,Data}=ssl:recv(S, 0), ssl:send(S, Data), ssl:close(L) end.
#Fun<erl_eval.20.52032458>
3> F().
```

Client:

```
→ openssl s_client -connect localhost:9234 -msg | grep -A3 'ServerKeyExchange'
depth=2 CN = erlangCA, OU = Erlang OTP, O = Ericsson AB, L = Stockholm, C = SE, emailAddress = peter@erix.ericsson.se
verify error:num=19:self signed certificate in certificate chain
verify return:0
<<< TLS 1.2 Handshake [length 019d], ServerKeyExchange
    0c 00 01 99 03 00 0e 91 04 02 9a b3 66 39 5f b9
    f6 30 e5 92 50 6e 7a 52 8e 0c f1 29 aa 11 3f f0
    c0 c2 31 15 0f ed d2 05 5a bd bb 80 c5 d9 03 94
```

`0c`: this is a ServerKeyExchange message (described in section 5.4 of RFC 4492)
`00 01 99`: of length 0x000199 bytes (409 bytes)
`03`: the curve type is "named_curve"
`00 0e`: the curve is `sect571r1` (curve identifiers are in section 5.1.1, identifier 0x000E is 14 in decimal)
---

Client preference config:

Server:

```
4> f(F), F= fun() -> {ok,L} = ssl:listen(9234, [{ecc_order, client_order}] ++ BaseConfig), 
4> {ok,S} = ssl:transport_accept(L), ok = ssl:ssl_accept(S),
4> {ok,Data}=ssl:recv(S, 0), ssl:send(S, Data), ssl:close(L) end.
#Fun<erl_eval.20.52032458>
5> F().
```

Client:

```
→ openssl s_client -connect localhost:9234 -msg | grep -A3 'ServerKeyExchange'
depth=2 CN = erlangCA, OU = Erlang OTP, O = Ericsson AB, L = Stockholm, C = SE, emailAddress = peter@erix.ericsson.se
verify error:num=19:self signed certificate in certificate chain
verify return:0
<<< TLS 1.2 Handshake [length 019d], ServerKeyExchange
    0c 00 01 99 03 00 0e 91 04 02 9a b3 66 39 5f b9
    f6 30 e5 92 50 6e 7a 52 8e 0c f1 29 aa 11 3f f0
    c0 c2 31 15 0f ed d2 05 5a bd bb 80 c5 d9 03 94
```

(same as before) 

`0c`: this is a ServerKeyExchange message (described in section 5.4 of RFC 4492)
`00 01 99`: of length 0x000199 bytes (409 bytes)
`03`: the curve type is "named_curve"
`00 0e`: the curve is `sect571r1` (curve identifiers are in section 5.1.1, identifier 0x000E is 14 in decimal)
---

Server with config that has nothing in common with the client:

Server:

```
6> f(F), F= fun() -> {ok,L} = ssl:listen(9234, [{ecc_order, [bacon]}] ++ BaseConfig), 
6> {ok,S} = ssl:transport_accept(L), ok = ssl:ssl_accept(S),
6> {ok,Data}=ssl:recv(S, 0), ssl:send(S, Data), ssl:close(L) end.
#Fun<erl_eval.20.52032458>
7> F().
```

Client:

```
→ openssl s_client -connect localhost:9234 -msg | grep -A3 'ServerKeyExchange'
140188496156320:error:14077410:SSL routines:SSL23_GET_SERVER_HELLO:sslv3 alert handshake failure:s23_clnt.c:770:
```

fails as expected
---

Server config picking another order, but common one:

Server:

```
8> f(F), F= fun() -> {ok,L} = ssl:listen(9234, [{ecc_order, [secp256r1]}] ++ BaseConfig), 
8> {ok,S} = ssl:transport_accept(L), ok = ssl:ssl_accept(S),
8> {ok,Data}=ssl:recv(S, 0), ssl:send(S, Data), ssl:close(L) end.
#Fun<erl_eval.20.52032458>
9> F().
```

Client:

```
→  openssl s_client -connect localhost:9234 -msg | grep -A3 'ServerKeyExchange'
depth=2 CN = erlangCA, OU = Erlang OTP, O = Ericsson AB, L = Stockholm, C = SE, emailAddress = peter@erix.ericsson.se
verify error:num=19:self signed certificate in certificate chain
verify return:0
<<< TLS 1.2 Handshake [length 014d], ServerKeyExchange
    0c 00 01 49 03 00 17 41 04 d9 51 be 9c 0e b6 8b
    fe ff db 2e 65 d6 83 33 1e 4d 12 c4 fe cf 54 9b
    28 d4 a3 c5 0f 68 a8 35 fa f6 78 71 d1 cd 38 5a
```

(same as before) 

`0c`: this is a ServerKeyExchange message (described in section 5.4 of RFC 4492)
`00 01 49`: of length 0x000149 bytes (329 bytes)
`03`: the curve type is "named_curve"
`00 17`: the curve is `secp256r1` (curve identifiers are in section 5.1.1, identifier 0x0017 is 23 in decimal)
---

Not exactly sure how to bring these tests within Erlang/OTP test suites.

ferd commented:
Other gotcha other than tests is whether we should have symmetry and allow clients to specify their priorities also. I'd like to add that after having figured tests out though, and there seems to be some tie-in with DTLS configs in there.

IngelaAndin commented:
I think the way to test it is to include the selected curve as an attribute that can be returned by ssl:connection_information/2 and test it in the same manner as when testing specific cipher suites
(that is done in ssl_basic_SUITE). That is  configure the client to only handle one curve and check that that one  is the one selected.  

I will get back to about other points after my teams PR meeting next week

ferd commented:
I've added tests to the PR along with the ssl:connection_info callback.

I also renamed the options from ecc_order to named_curve_order. The terminology
is a bit vague, but there's plenty of unnamed curves; named ones appear
to only be working with servers accepting ECDHE_ECDSA.

To make the tests work, a value had to be added to
connection_information for the SSL value. The value requires specific
alterations to work, since for example, a server will *always* carry a
named curve parameter in its session state, even if it didn't use it
(the selection is done regardless of usage, as long as the client submitted
the named curve extension).

The client on the other end, never actually noted down the named curve
that was used (if any), and hooks had to be added in the decoding of the
handshake to carry and be able to expose the information.

Based on the presence of ECDHE_ECDSA, the connection_information value
for named curves is only shown if it was used at all.

The session data has some risk of being stale after a renegotiation
where the initial handshake was performed with an ECDHE_ECDSA
ciphersuite, but the latter one was done without it. In such cases, the
negotiated cipher suite should still prevent the connection_information
from being wrong.

I've let things sit in two commits for now for ease of reviewing, but can merge
them later or on-demand. I have no idea how to expose other ECC curves
than the named ones, and this does not quite fit into the idea of just allowing
named curve prioritization I initially wanted. There's ways to extend this to
other types of curves, but I haven't looked into it right. The design may however
change if that were to be supported (I think the `ecc` field is only used for named
curves, non-named ones seem to find or generate params from elsewhere)

OTP-Maintainer commented:
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

IngelaAndin commented:
This feels more complex then I first thought. Here are some
comments after our PR-meeting. 

When it comes to preferring the server order and being able to
specify which algorithms the server will support we think that it
probably should be two options. It could be possible that you
want to limit the supported algorithms instead of using the
default but still honour the clients preference, which maybe should
be the default.

The current implementation honers the servers order but check that it
is in the clients list. If you see to how cipher suites are handled I
think it would be logical that the clients preference is
preferred. Also the RFC is talking about the client list being
ordered. However I think the following passage of the RFC is a bit
vague on this point.

"A server that receives a ClientHello containing one or both of these
extensions MUST use the client's enumerated capabilities to guide its
selection of an appropriate cipher suite."

If you look at the server option honor_cipher_order it feel similar
to the situation we have now. 

Currently the ssl application does not support arbitrary curves
even though most of the support for it is in place, there is
no way to configure it. As we read the RFC 4492 there are to
special enum values to in section 5.1.1:

arbitrary_explicit_prime_curves(0xFF01),
arbitrary_explicit_char2_curves(0xFF02),

to signal use of these curves. So maybe ECC is a good name name
although only named curves are supported at the moment. I suppose
it would be up to the server to choose an arbitrary curve and that
might require some configuration as I do not think you want it
to totally arbitrary. Probably it would be similar to DH-configuration.

To reduce the code your functions select_from_client and select_from_server ought to
be able to be calls to one function with reversed arguments.

Thoughts?

ferd commented:
Just so I make sure I understand:

- break up the current implementation into two options: `ecc` and `honor_ecc_order` 
- `ecc` allows the specification of the supported named curves, in their preferred order, and may eventually support more values for explicit primes and so on.
- `honor_ecc_order` allows to choose whether to honor the elliptic curves (currently only named) order from the client, the way it is done for ciphersuites

Am I also right in understanding that it is okay to only implement this for the server for now, or should it also be done for the client? Is the testing strategy needing to change since we'd be moving away from the current implementation (which uses hardcoded values)? Should the API also be expanded in similar ways (i.e. should there be an `ssl:ecc()` and `ssl:ecc(erlang)` functions?) or this can be kept internal just for the sake of filtering the configuration early on (currently that code lives in `tls_v1.erl` and some `public_key` module)?

IngelaAndin commented:
In the long run we will want it for the client as well as the server. So I think we need to think about how it shall be done for the client even if you should decide to only contribute the server side. 
 
ecc option should exist for both server  and client and define the set of supported curves in the preferred order. There should exist a default value corresponding to the current hard coded list.

The server could always prefer the client order unless the server side option honor_ecc_order is set
to true, and it should be set to false by default.  This is not fully backwards compatible, but the option makes it possible to make compatible. And I do not think that it would be a notable change. 

I was hoping you had a convincing argument to make it behave like this as I think the RFC is not  really clear and it is not obvious the current implementation is flawed, even if I am inclined to reason like you did.  

Yes I  think it would be nice to have an ssl:ecc function as TLS  RFCs have their own ECC definitions
and it does not map one to one with crypto:ec_curves().

ferd commented:
Alright, I'll try to get started with that vision and implementation

IngelaAndin commented:
FYI the tls_v1 module has encryption related help functions to handle TLS versions (TLS 1.0, TLS 1.1 and TLS 1.2) as opposed to  SSL version  (SSL 3.0) that is handled by the ssl_v3 module. Actually, there also exists a ssl_v2 module to handle compatibility with clients that offer SSL 2.0 but support  higher versions. The latter stuff is no longer default supported, but still available.

Also, PR does not have to handle arbitrary curves. I just want it to be extendable to do so some day.

ferd commented:
The latest commit adds to the previous ones by changing the way
ECC selection is done by adding two options:

- `eccs`, shared by clients and servers alike, allows the specification
  of the supported named curves, in their preferred order, and may
  eventually support more values for explicit primes and so on.
- `honor_ecc_order`, a server-only option, is similar to
  `honor_cipher_order` and will, by default let the server pick the
  client-preferred ECC, and otherwise pick the server-preferred one.

The default value for `eccs` is the same as before, although the
server-chosen ECC now defaults to the client rather than previous
choice.

A function `ssl:eccs()` has been added that returns the highest
supported ECCs for the library.

It turned out to be easier than I thought to allow customizing ECCs
for the client, so I bundled it and added tests for that.

The only thing I'm still not 100% sure is the default curve selection done
by `ssl:eccs()` being done by using the `tls_v1` module hardcoded. It sounds
like it should be possible to just add other modules supporting them (`dtls`) 
and then get the ball rolling, but I'm not sure what preference order
should be given between both, and I punted on it for now given currently
`dtls` just wraps `tls_v1`.

The option name `eccs` was chosen because the extracted and chosen ECC
is under the name `ecc`; if the option was named `ecc` also, then the selected
ECC would overwrite or be overwritten in `ssl:connection_info` by the configured
set of named curves.

ferd commented:
I think the CI check failed due to a bad XML tag in docs. I fixed it, amended, and pushed, but the job hasn't run again for it yet.

IngelaAndin commented:
Thanks we will have an other look. DTLS  is specified as a diff towards TLS and that is how I want  to 
organise the code, although there will be two separate gen_statem processes that share  code.

IngelaAndin commented:
ssl:eccs/0 is meant to be an API function and then it needs documentation. I see what you mean by hardcoding tls_v1. The thing to do is to add an ssl_cipher:eccs(Version) like there is an ssl_cipher:suites/1.

```
> suites({3, 0}) ->
>     ssl_v3:suites();
> suites({3, N}) ->
>     tls_v1:suites(N).
> 
> 
```
DTLS is handled in many places by converting its version to corresponding TLS version if there is
no special only DTLS handling.

ferd commented:
I've just added a fixed eccs/0-1 set of functions, with docs and tests.

OTP-Maintainer commented:
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

IngelaAndin commented:
We do not need the clause 

`+eccs({2,0}) ->
 +    [];

as we have never supported SSL-2.0, we only have the possibility to parse  SSL-2.0 hellos
in order to interop with clients that offer SSL-2.0 but also higher version.

ferd commented:
Fixed.

IngelaAndin commented:
New code included in the test build.

OTP-Maintainer commented:
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

IngelaAndin commented:
I have looked at test results now and there are a few problems. The PR generates 17 dialyzer errors.
I can send them to you in an email. 

On old versions of openssl like OpenSSL 0.9.8k ssl:eccs()  will return an empty list failing some of
the test cases.

Also on machines with "OpenSSL 1.0.1i-fips 6 Aug 2014"  we get

```
=== Ended at 2016-10-31 05:47:44
=== Location: [{ssl_ECC_SUITE,ecc_test,361},
              {test_server,ts_tc,1529},
              {test_server,run_test_case_eval1,1045},
              {test_server,run_test_case_eval,977}]
=== Reason: {failed,
                     {test_case_failed,
                         {{expected,{<0.2606.0>,ok}},
                          {expected,{<0.2604.0>,ok}},
                          {got,
                              {<0.2604.0>,
                               {error,server,sect571r1,
                                   {ecc,{named_curve,secp256r1}}}}}}}}

```
I also suspect there might be an ECC handling bug unrelated to your PR.  I will include some details
in the dialyzer email.

ferd commented:
I have changed things so that Dialyzer warnings should be gone, and have made the ECC test runs conditional to whether its configured curves are supported in each individual test, since I suspect this may be the failure here.

IngelaAndin commented:
Great back to the test loop then :)

IngelaAndin commented:
Looking better but I still ssl_basic_SUITE eccs fails. I think it has to be conditioned too.

```
** CT Error Notification 2016-11-02 05:09:03.702 ***
ssl_basic_SUITE:eccs failed on line 1514
Reason: {badmatch,[]}

Full error description and stacktrace

=== Ended at 2016-11-02 05:09:03
=== Location: [{ssl_basic_SUITE,eccs,1514},
              {test_server,ts_tc,1529},
              {test_server,run_test_case_eval1,1045},
              {test_server,run_test_case_eval,977}]
=== Reason: no match of right hand side value []
  in function  ssl_basic_SUITE:eccs/1 (ssl_basic_SUITE.erl, line 1514)
  in call from test_server:ts_tc/3 (test_server.erl, line 1529)
  in call from test_server:run_test_case_eval1/6 (test_server.erl, line 1045)
  in call from test_server:run_test_case_eval/9 (test_server.erl, line 977)

```

ferd commented:
Ah yeah that one slipped under the radar. Patched it in. Do you want me to also rebase all the commits for this PR into a single one at this point?

IngelaAndin commented:
Yes please do.

ferd commented:
All clear.

OTP-Maintainer commented:
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------
